### PR TITLE
[Reformer] - Cache hidden states and buckets to speed up inference

### PR DESCRIPTION
**Pull request by patrickvonplaten**
As discussed with the authors in Reformer we cache the hidden states and buckets to speed up inference for language generation. Caching only the hidden states can save at least twice the memory versus caching the key and value output vectors (often more since the `hidden_size` can be smaller than key and query projections). 

The idea is to only recompute key and value projections within the same chunk so that the output will be equal.

- [x] Implement caching mechanism for Local Attention
- [x] Implement caching mechanism for LSH Attention
- [x] Add test
- [x] Add logic to generation
- [x] Refactoring and better naming

=> This results in a 10x speed up when generating up to 1000 tokens.

**Some fun trials trying to make Reformer generate 16000 tokens for its own Wikipedia article [here](https://colab.research.google.com/drive/1Oao8vBtDkz6v1E1efUhTlug5uuxC_BnE?usp=sharing).**

*Review*:
Not urgent. Merging this PR can wait a bit. Added a couple of tests to make sure caching gives same output as non-caching.

### CODE REVIEW

File: src/transformers/modeling_xlnet.py [hunk 1]
```diff
@@ -600,7 +600,7 @@ class XLNetModelOutput(ModelOutput):
 @dataclass
 class XLNetLMHeadModelOutput(ModelOutput):
     """
-    Output type of :class:`~transformers.XLNetModel`.
+    Output type of :class:`~transformers.XLNetLMHeadModel`.
```
Comment (patrickvonplaten) [inline src/transformers/modeling_xlnet.py:603]: @sgugger - renamed here. Think this was leftover from copy paste

File: src/transformers/modeling_xlnet.py [hunk 2]
```diff
@@ -637,7 +637,7 @@ class XLNetLMHeadModelOutput(ModelOutput):
 @dataclass
 class XLNetForSequenceClassificationOutput(ModelOutput):
     """
-    Base class for outputs of sentence classification models.
+    Output type of :class:`~transformers.XLNetForSequenceClassification`.
```
Comment (patrickvonplaten) [inline src/transformers/modeling_xlnet.py:640]: @sgugger - renamed here. Think this was leftover from copy paste

File: src/transformers/modeling_xlnet.py [hunk 3]
```diff
@@ -671,7 +671,7 @@ class XLNetForSequenceClassificationOutput(ModelOutput):
 @dataclass
 class XLNetForTokenClassificationOutput(ModelOutput):
     """
-    Base class for outputs of token classification models.
+    Output type of :class:`~transformers.XLNetForTokenClassificationOutput`.
```
Comment (patrickvonplaten) [inline src/transformers/modeling_xlnet.py:674]: @sgugger - renamed here. Think this was leftover from copy paste

File: tests/test_modeling_reformer.py [hunk 4]
```diff
@@ -440,6 +468,35 @@ def create_and_check_longformer_for_question_answering(self, config, input_ids,
...
+        )
+        outputs_without_cache = model(input_ids)[0][:, -1]
+
+        # select random slice idx
+        random_slice_idx = torch.randint(outputs_without_cache.shape[-1], (1, 1), device=torch_device).item()
+
+        # outputs should be similar within range
+        self.parent.assertTrue(
+            torch.allclose(
+                outputs_with_cache[:, 0, random_slice_idx], outputs_without_cache[:, random_slice_idx], atol=1e-2
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:496]: only `1e-2` because caching leads to minimal different results due to multiple masking (self mask + causal mask) and different padding for incremental / non-incremental

File: tests/test_modeling_reformer.py [hunk 5]
```diff
@@ -1049,8 +1120,23 @@ def test_pretrained_generate_crime_and_punish(self):
...
             "A few months later state expression in his ideas, at the first entrance. He was positively for an inst",
         )
+
+    @slow
+    def test_pretrained_generate_use_cache_equality(self):
+        model = ReformerModelWithLMHead.from_pretrained("google/reformer-crime-and-punishment").to(torch_device)
+        tokenizer = ReformerTokenizer.from_pretrained("google/reformer-crime-and-punishment")
+        model.eval()
+        input_ids = tokenizer.encode("A few months later", return_tensors="pt").to(torch_device)
+        output_ids_with_cache = model.generate(input_ids, max_length=130, num_hashes=8, use_cache=False)
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:1136]: need `num_hashes=8` to ensure equal output.

File: tests/test_modeling_reformer.py [hunk 6]
```diff
@@ -348,7 +357,9 @@ def create_and_check_reformer_feed_backward_chunking(self, config, input_ids, in
         config.chunk_size_feed_forward = 1
 
         torch.manual_seed(0)
-        model = ReformerModelWithLMHead(config=config)
+
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:360]:
```suggestion
```

File: tests/test_modeling_reformer.py [hunk 7]
```diff
@@ -348,7 +357,9 @@ def create_and_check_reformer_feed_backward_chunking(self, config, input_ids, in
...
+        model = ReformerForMaskedLM(config=config)
+
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:362]:
```suggestion
```

File: tests/test_modeling_reformer.py [hunk 8]
```diff
@@ -332,9 +337,13 @@ def create_and_check_reformer_feed_backward_chunking(self, config, input_ids, in
         config.hidden_dropout_prob = 0
         config.local_attention_probs_dropout_prob = 0
         config.lsh_attention_probs_dropout_prob = 0
+        config.lsh_num_chunks_after = 1
+        config.is_decoder = False
 
         torch.manual_seed(0)
-        model = ReformerModelWithLMHead(config=config)
+
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:344]:
```suggestion
```

File: tests/test_modeling_reformer.py [hunk 9]
```diff
@@ -332,9 +337,13 @@ def create_and_check_reformer_feed_backward_chunking(self, config, input_ids, in
...
+        model = ReformerForMaskedLM(config=config)
+
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:346]:
```suggestion
```

File: tests/test_modeling_reformer.py [hunk 10]
```diff
@@ -193,17 +193,22 @@ def create_and_check_reformer_model(self, config, input_ids, input_mask, choice_
         )
 
     def create_and_check_reformer_model_with_lm_backward(self, config, input_ids, input_mask, choice_labels):
-        model = ReformerModelWithLMHead(config=config)
+        config.is_decoder = False
+        config.lsh_num_chunks_after = 1
+        model = ReformerForMaskedLM(config=config)
+
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:199]:
```suggestion
```

File: src/transformers/modeling_reformer.py [hunk 11]
```diff
@@ -372,8 +481,12 @@ def forward(
             value_vectors.shape[-1], self.attention_head_size
         )
 
+        do_standard_self_attention = (sequence_length <= self.chunk_length) or (
+            use_cache and cached_hidden_states_and_buckets[1] is not None
+        )
+
```
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:487]:
```suggestion
```

File: src/transformers/modeling_reformer.py [hunk 12]
```diff
@@ -80,7 +83,18 @@ def mish(x):
...
     "ReformerBackwardOutput", ["attn_output", "hidden_states", "grad_attn_output", "grad_hidden_states"]
 )
-ReformerEncoderOutput = namedtuple("ReformerEncoderOutput", ["hidden_states", "all_hidden_states", "all_attentions"])
+ReformerEncoderOutput = namedtuple(
+    "ReformerEncoderOutput", ["hidden_states", "all_hidden_states", "all_attentions", "past_buckets_states"],
+)
+
+
+def _stable_argsort(vector, dim):
+    # this function scales the vector so that torch.argsort is stable.
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:92]: what does stable mean here? inconsistent tiebreaking? just curious.

File: src/transformers/modeling_reformer.py [hunk 13]
```diff
@@ -100,6 +114,23 @@ def _get_least_common_mult_chunk_len(config):
         )
 
 
+def _get_min_chunk_len(config):
+    attn_types = config.attn_layers
+    attn_types_set = set(attn_types)
+    if len(attn_types_set) == 1 and attn_types[0] == "lsh":
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:120]: (NITCITY)
you can do set equality I think. You can also type hint instead of including type in varname.
```python
if attn_types == {'lsh'}:
```

File: src/transformers/modeling_reformer.py [hunk 14]
```diff
@@ -171,15 +202,23 @@ def forward(self, position_ids):
...
             position_encodings = torch.cat(
                 [weight[:, :required_pos_encodings_columns] for weight in broadcasted_weights], dim=-1
             )
-            position_encodings = torch.reshape(position_encodings, (batch_size, -1, position_encodings.shape[-1]))[
-                :, :sequence_length
-            ]
+            position_encodings = torch.reshape(position_encodings, (batch_size, -1, position_encodings.shape[-1]))
+
+            # select correct position encodings
+            position_encodings = torch.cat(
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:215]: can you batch this somehow
`position_encodings.index_select(1, position_ids)` maybe?

File: src/transformers/modeling_reformer.py [hunk 15]
```diff
@@ -349,18 +392,73 @@ def forward(
         # num hashes can optionally be overwritten by user
         num_hashes = num_hashes if num_hashes is not None else self.num_hashes
 
-        # project hidden_states to query_key and value
-        query_key_vectors = self.query_key(hidden_states)
-        value_vectors = self.value(hidden_states)
+        do_cached_attention = use_cache and past_buckets_states[1] is not None
+
+        # check if cache shall be used and that hidden states are already cached
+        if do_cached_attention:
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:398]: why not always use the cache?

File: src/transformers/modeling_reformer.py [hunk 16]
```diff
@@ -349,18 +392,73 @@ def forward(
...
+            assert (
+                sequence_length == 1
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:400]: wonderful assert

File: src/transformers/modeling_reformer.py [hunk 17]
```diff
@@ -466,9 +580,28 @@ def forward(
...
+
         return LSHSelfAttentionOutput(hidden_states=out_vectors, attention_probs=attention_probs, buckets=buckets)
 
-    def _hash_vectors(self, vectors, num_hashes, attention_mask):
+    def _query_per_attn_head(self, hidden_states):
+        per_head_query_key = self.query_key.weight.reshape(
+            self.num_attention_heads, self.attention_head_size, self.hidden_size
+        ).transpose(-2, -1)
+        # only relevant for inference and no bias => we can use einsum here
+        query_key_vectors = torch.einsum("balh,ahr->balr", hidden_states, per_head_query_key)
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:593]: blahhr

what are these letters!

File: src/transformers/modeling_reformer.py [hunk 18]
```diff
@@ -600,26 +722,37 @@ def _attend(
...
         # get logits and dots
+        # (BS, NumAttn, NumHash x NumChunk, Chunk_L x Hidden),(BS, NumAttn, NumHash x NumChunk, Chunk_L * (Num_bef + Num_aft + 1) x Hidden) -> (BS, NumAttn, NumHash x NumChunk, Chunk_L, Chunk_L * (1 + Num_bef + Num_aft))
         query_key_dots = torch.matmul(query_vectors, key_vectors.transpose(-1, -2))
 
         # free memory
         del query_vectors, key_vectors
 
         # if chunked attention split bucket idxs to query and key
-        if self.chunk_length < sequence_length:
+        if not do_standard_self_attention:
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:741]: do we hit all these cases? Do we need to?

File: src/transformers/modeling_reformer.py [hunk 19]
```diff
@@ -715,6 +854,102 @@ def _compute_attn_mask(self, query_indices, key_indices, attention_mask, query_k
...
+        self, query_vectors, attention_mask, num_hashes, hidden_states, past_states, past_buckets
+    ):
+        # concat hidden states
+        hidden_states = torch.cat([past_states, hidden_states], dim=1)
+
+        # batch_size hidden
+        batch_size = hidden_states.shape[0]
+        sequence_length = hidden_states.shape[1]
+
+        # check if cached buckets include pad bucket
```
Comment (sshleifer) [inline src/transformers/modeling_reformer.py:867]: good comment. I get it (more) now

File: tests/test_modeling_reformer.py (hunk 5 above)
Comment (sshleifer) [reply tests/test_modeling_reformer.py:1129]: why?

consider testing some bad inputs and making sure you raise the errors you want.

This is a lot of code for one test, albeit a strong one. Do you need a big model to run this test though?

**Review by sshleifer — APPROVED** [review]

File: tests/test_modeling_reformer.py (hunk 5 above)
Comment (patrickvonplaten) [reply tests/test_modeling_reformer.py:1129]: Yeah, good question. So in general using the `cache` vs. not using the cache is never 100% the same because when one does not use the cache, the values are masked by -10^5 which is not the same as not calculating them at all. That's why all our `use_cache == no_use_cache` tests compare on 10e-3 level because at 10e-5 they would already fail (tested this with T5). 

In the case of Reformer, it is even trickier because there is a `self_mask` (masked the token of the same position) and a `causal_mask` which are different (1e-5 for `self_mask` and 1e-9 for `causal_mask`) because there are cases where all tokens have 1e-9 and only the first position has 1e-5 (this happens for the first token of the sequence for example) => therefore the values are much more sensible to removing the masks completely when doing cached inference.

`num_hashes` remedies this because it averages over `num_hashes` query-key calculations so the results become more stable. No need to run this test on a big model though.

File: src/transformers/modeling_reformer.py (hunk 12 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_reformer.py:92]: yeah inconsistent tiebreaking I think it is. `e.g.` stable argmax wolud solt [1, 0, 0] always as [1, 2, 0] but unstable which is the only option for `torch.sort` can have either [1, 2, 0] or [2, 1, 0] => this function uses a trick to make sure it's always stable.

File: src/transformers/modeling_reformer.py (hunk 14 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_reformer.py:215]: tried it - doesn't work sadly :-/

File: src/transformers/modeling_reformer.py (hunk 15 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_reformer.py:398]: Even if `use_cache=True`, the first inference step will have `past_buckets_states == None` => so it should not do `do_cached_attention`.

File: src/transformers/modeling_reformer.py (hunk 17 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_reformer.py:593]: haha -> b = `batch_size` a = `num_attention_heads`, l = `seq_len`, h = `hidden_size`, r = `projection_size`

File: src/transformers/modeling_reformer.py (hunk 18 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_reformer.py:741]: Yes lot's of weird special cases for Reformer caching

**Comment by patrickvonplaten** [comment]
@LysandreJik @sgugger - when you guys are back, feel free to merge to PR or we can wait until I'm back - not too urgent this PR. LGTM to merge.

**Comment by patrickvonplaten** [comment]
Okey merging this. 

@sgugger @LysandreJik @sshleifer -> checked caching extensively and added multiple tests for equality. 
Would be nice if you can tag me for possible future issues related to this caching mechanism.

File: src/transformers/modeling_xlnet.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_xlnet.py:603]: Yes, thanks for fixing!
